src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,416 @@
+
+            return result.ToString();
+        }
+
+        /// <returns>Positive length or negative value for invalid name</returns>
+        internal static int GetFullTypeNameLength(ReadOnlySpan<char> input, out bool isNestedType)
+        {
+            isNestedType = false;
+
+            // NET 6+ guarantees that MemoryExtensions.IndexOfAny has worst-case complexity
+            // O(m * i) if a match is found, or O(m * n) if a match is not found, where:
+            //   i := index of match position
+            //   m := number of needles
+            //   n := length of search space (haystack)
+            //
+            // Downlevel versions of .NET do not make this guarantee, instead having a
+            // worst-case complexity of O(m * n) even if a match occurs at the beginning of
+            // the search space. Since we're running this in a loop over untrusted user
+            // input, that makes the total loop complexity potentially O(m * n^2), where
+            // 'n' is adversary-controlled. To avoid DoS issues here, we'll loop manually.
+
+#if NET8_0_OR_GREATER
+            int offset = input.IndexOfAny(_endOfFullTypeNameDelimitersSearchValues);
+            if (offset < 0)
+            {
+                return input.Length; // no type name end chars were found, the whole input is the type name
+            }
+
+            if (input[offset] == EscapeCharacter) // this is very rare (IL Emit or pure IL)
+            {
+                offset = GetUnescapedOffset(input, startOffset: offset); // this is slower, but very rare so acceptable
+            }
+#else
+            int offset = GetUnescapedOffset(input, startOffset: 0);
+#endif
+            isNestedType = offset > 0 && offset < input.Length && input[offset] == '+';
+            return offset;
+
+            static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
+            {
+                int offset = startOffset;
+                for (; offset < input.Length; offset++)
+                {
+                    char c = input[offset];
+                    if (c == EscapeCharacter)
+                    {
+                        offset++; // skip the escaped char
+
+                        if (offset == input.Length || // invalid name that ends with escape character
+                            !NeedsEscaping(input[offset])) // invalid name, escapes a char that does not need escaping
+                        {
+                            return -1;
+                        }
+                    }
+                    else if (NeedsEscaping(c))
+                    {
+                        break;
+                    }
+                }
+                return offset;
+            }
+
+            static bool NeedsEscaping(char c) => c is '[' or ']' or '&' or '*' or ',' or '+' or EscapeCharacter;
+        }
+
+        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName)
```

Comment:
What does `GetName(@"Hello\\.There")` return?

Reply:
> What does `GetName(@"Hello\\.There")` return?

It returns `Hello\\.There` and I have test cases for similar scenarios: 

https://github.com/dotnet/runtime/blob/04731fba62eda36df59a47fd9ca9e46ef351cf1f/src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameParserHelpersTests.cs#L69-L78